R/bundlePackagePackrat.R

```diff
@@ -0,0 +1,200 @@
+  })
+
+  records
+}
+
+standardizeRecords <- function(records, repos) {
+  availablePackages <- availablePackages(repos)
+  repos <- standardizeRepos(repos)
+
+  rows <- lapply(seq_len(nrow(records)), function(i) {
+    standardizePackratPackage(records[i, ], availablePackages, repos = repos)
+  })
+  rows <- lapply(rows, as.data.frame, stringsAsFactors = FALSE)
+  rbind_fill(rows, c("Source", "Repository"))
+}
+
+standardizeRepos <- function(repos) {
+  # Ensure that each repository has a unique name
+  names(repos) <- ifelse(
+    names2(repos) == "",
+    paste0("repo_", seq_along(repos)),
+    names2(repos)
+  )
+
+  # And strip trailing /
+  repos <- gsub("/$", "", repos)
+
+  repos
+}
+
+standardizePackratPackage <- function(record, availablePackages, repos = character()) {
+  pkg <- record$Package
+  source <- record$Source
+
+  # source types are defined by packrat:
+  # https://github.com/rstudio/packrat/blob/v0.9.0/R/pkg.R#L328
+  if (source %in% c("github", "gitlab", "bitbucket")) {
+    # SCM information is recorded elsewhere
+    repository <- NA_character_
+  } else if (source == "source") {
+    # can't install source packages elsewhere
+    repository <- NA_character_
+    source <- NA_character_
+  } else if (source == "CustomCRANLikeRepository") {
+    # Package was installed from source, but packrat guessed it was installed
+    # from a known repo. We need to check that it's not actually a dev version
+    if (isDevVersion(record, availablePackages)) {
+      repository <- NA_character_
+      source <- NA_character_
+    } else {
+      repository <- findRepoUrl(pkg, availablePackages)
+      source <- findRepoName(repository, repos)
+    }
+  } else if (source %in% c("CRAN", "Bioconductor")) {
+    # shinyapps & posit.cloud will ignore, but connect will use (unless admin
+    # has set up an override)
+    repository <- findRepoUrl(pkg, availablePackages)
+  } else {
+    # Installed from custom repository. Find URL from available.packages()
+    # and then name from repos.
+    repository <- findRepoUrl(pkg, availablePackages)
+    source <- findRepoName(repository, repos)
+  }
+  list(Source = source, Repository = repository)
+}
+
+findRepoName <- function(repository, repos) {
+  idx <- match(repository, repos)
+  names(repos)[idx]
+}
+
+findRepoUrl <- function(pkg, availablePackages) {
+  if (pkg %in% rownames(availablePackages)) {
+    repo <- availablePackages[pkg, "Repository"]
+    # Strip `/src/contrib/*` from package repository: `contrib.url()`
+    # adds /src/contrib, and RSPM adds additional directories
+    gsub("/src/contrib.*$", "", repo)
+  } else {
+    NA_character_
+  }
+}
+
+isDevVersion <- function(record, availablePackages) {
+  if (!record$Package %in% rownames(availablePackages)) {
+    return(FALSE)
+  }
+
+  local_version <- record$Version
+  repo_version <- availablePackages[record$Package, "Version"]
+
+  package_version(local_version) > package_version(repo_version)
```

Comment:
Should we also detect whether the trailing version component is "large"? In case a local version is 1.0.0-9000 but CRAN has 1.0.1?

Reply:
That's not a version used by every package, so I worry that would dilute the impact of this change.